File: Tree/BST/bst.c
#include <stdio.h>
#include <stdlib.h>
typedef int val_t;
typedef int msize_t;

typedef struct Node_t
{
		struct Node_t *left;
		struct Node_t *right;
		val_t value;
}node_t;

typedef struct Search_t
{
		struct Node_t *curr;
		struct Node_t *parent;
}search_t;

typedef struct QNode_t
{
		struct Node_t *value;
		struct QNode_t *next;
		struct QNode_t *prev;
}qnode_t;

typedef struct Queue_t
{
		struct QNode_t *head;
		struct QNode_t *tail;
		msize_t length;
}queue_t;

queue_t *queue_global;
int helper2_2(node_t *);

int max(int a, int b)
{
		return(a > b ? a:b);
}

int min(int a, int b)
{
		return(a < b ? a:b);
}

queue_t *init_queue(void)
{
		queue_t *q = malloc(sizeof(queue_t));
		q->head = NULL;
		q->tail = NULL;
		q->length = 0;
		return(q);
}


void *enqueue(queue_t *queue, node_t *node)
{
		qnode_t *new_qnode = malloc(sizeof(qnode_t));
		new_qnode->value = node;
		new_qnode->prev = NULL;
		new_qnode->next = NULL;
		if (queue->length == 0)
		{
				queue->head = queue->tail = new_qnode;
				queue->length++;
		}
		else
		{
				queue->tail->next = new_qnode;
				new_qnode->prev = queue->tail;
				queue->tail = new_qnode;
				queue->length++;
		}
}

node_t *dequeue(queue_t *queue)
{
		//printf("dequeue: %d\n", queue->head->value->value);
		if (queue->length <= 0)
				return(NULL);
		qnode_t *first_q = queue->head;
		node_t *ret_node = first_q->value;
		if (queue->length == 1)
				queue->head = queue->tail = NULL;
		else
		{
				queue->head = first_q->next;
				queue->head->prev = NULL;
		}
		queue->length--;
		free(first_q);
		return(ret_node);
}

		

node_t *getleft(node_t *node)
{
		return(node->left);
}

node_t *getright(node_t *node)
{
		return(node->right);
}

val_t getvalue(node_t *node)
{
		return(node->value);
}

node_t *create_node(val_t value)
{
		node_t *new_node = malloc(sizeof(node_t));
		new_node->left = NULL;
		new_node->right = NULL;
		new_node->value = value;
		return(new_node);
}

int insert(node_t *tree, val_t value)
{
		node_t *new_node = create_node(value);
		node_t *parent;
		val_t curr_val;
		while (tree != NULL)
		{
				if ((curr_val = getvalue(tree)) == value)
						return(-1);
				else if (value < curr_val)
					{ parent = tree; tree = getleft(tree); }
				else
					{ parent = tree; tree = getright(tree);}
		}
		if (value < curr_val)
				parent->left = new_node;
		else
				parent->right = new_node;
		return(1);
}

struct Search_t search(node_t *tree, val_t value)
{
		val_t curr_val;
		node_t *parent = NULL;
		struct Search_t result;
		while (tree != NULL)
		{
				curr_val = getvalue(tree);
				if (value == curr_val)
						{ result.parent = parent; result.curr = tree; return(result);}
				else if (value < curr_val)
						{ parent = tree; tree = getleft(tree);}
				else
						{ parent = tree; tree = getright(tree);}
		}
		result.parent = NULL;
		result.curr = NULL;
		return(result);
}


struct Search_t successor(node_t *tree)
{
		node_t *parent = tree;
		tree = getright(tree);
		node_t *succ = tree;
		struct Search_t result;
		while (tree != NULL)
		{
				succ = tree; 
				tree = getleft(tree); 
				if (tree != NULL && getleft(tree) != NULL)
						parent = succ;
			
		}
		result.parent = parent;
		result.curr = succ;
		return(result);
}


val_t delete(node_t *tree, val_t value)
{
		node_t *delnode, *left, *right, *parent, *succ, *succ_parent;
		struct Search_t result = search(tree, value);
		val_t del_val;
		if ((delnode = result.curr) == NULL)
				return(-1);
		parent = result.parent;
		del_val = getvalue(delnode);
		left = getleft(delnode);
		right = getright(delnode);
		if (left == NULL && right == NULL)
		{
				if (getleft(parent) == delnode)
						parent->left = NULL;
				else
						parent->right = NULL;
				free(delnode);
		}
		else if (!(left != NULL && right != NULL)) // one child
		{

				if (left != NULL)
						parent->left = left;
				else
						parent->right = right;
				free(delnode);
		}
		else // two child
		{
				result = successor(delnode);
				succ_parent = result.parent;
				succ = result.curr;
				delnode->value = getvalue(succ);
				if (getright(succ) != NULL)
				{
						succ_parent->right = getright(succ);
				}
				free(succ);
		}
		return(del_val);
}


void preorder_traversal(node_t *tree)
{
		if (tree != NULL)
		{
				printf("%d\t", getvalue(tree));
				preorder_traversal(getleft(tree));
				preorder_traversal(getright(tree));
		}
}

void postorder_traversal(node_t *tree)
{
		if (tree != NULL)
		{
				postorder_traversal(getleft(tree));
				postorder_traversal(getright(tree));
				printf("%d\t", getvalue(tree));
		}
}

void inorder_traversal(node_t *tree)
{
		if (tree != NULL)
		{
				inorder_traversal(getleft(tree));
				printf("%d\t", getvalue(tree));
				inorder_traversal(getright(tree));
		}
}

void levelorder_helper(void)
{
		node_t *tree;
		while ((tree = dequeue(queue_global)) != NULL)
		{
				printf("%d\t", getvalue(tree));
				if (getleft(tree) != NULL)
						enqueue(queue_global, getleft(tree));
				if (getright(tree) != NULL)
						enqueue(queue_global, getright(tree));
		}
}

void levelorder_traversal(node_t *tree)
{
		if (tree != NULL)
		{
				enqueue(queue_global, tree);
				levelorder_helper();
		}
}

int find_height(node_t *tree)
{
		// return the count of edges to the leaf
		if (tree == NULL)
				return(-1);
		return(max(find_height(tree->left), find_height(tree->right)) + 1);
}

int find_node(node_t *tree)
{
		return(find_height(tree) - 1);
}

int depth_helper(node_t *node, val_t value, int depth)
{
		if (node == NULL)
				return(-1);
		if (getvalue(node) == value)
				return(depth);
		return(max(depth_helper(getleft(node), value, depth+1), depth_helper(getright(node), value, depth+1)));
}


int find_depth(node_t *tree, val_t value)
{
		return(depth_helper(tree, value, 0));
}

int isbinaryhelper(node_t *node, val_t min_val, val_t max_val)
{
		if (node == NULL)
				return(1);
		if (getvalue(node) > min_val && getvalue(node) < max_val)
				return(min(isbinaryhelper(getleft(node), min_val, getvalue(node)), isbinaryhelper(getright(node), getvalue(node), max_val)));
		return(-1);
}

int isbinaryhelper2_inorderhelper(node_t *node)
{
		// must be called from isbinarthelper2, or error will occur
		static val_t prev = 1 << 31;
		if (getvalue(node) <= prev)
				return(-1);
		prev = getvalue(node);
		return(1);
}

int isbinarytree2(node_t *node)
{
		if (node != NULL)
		{
				if (isbinarytree2(getleft(node)) > 0)
						if (isbinaryhelper2_inorderhelper(node) > 0)
							if (isbinarytree2(getright(node)) > 0)
									return(1);
				return(-1);
		}
		return(1);
}




int isbinarytree(node_t *tree)
{
		return(isbinaryhelper(tree, 1<<31, 1<<31 ^ -1));
}

void delete_tree(node_t *tree)
{
		if (tree != NULL)
		{
				delete_tree(getleft(tree));
				delete_tree(getright(tree));
				free(tree);
		}
}



int main(int argc, char *argv[])
{
		/* 
		   5
		 /  \
		3   10
	   /     \
	   2     15
	  /
     1
	 */
		queue_global = init_queue();
		node_t *tree = create_node(5);
		insert(tree, 10);
		insert(tree, 15);
		insert(tree, 3);
		insert(tree, 2);
		insert(tree, 1);
		delete(tree, 1);
		delete(tree, 15);
		insert(tree, 1);
		insert(tree, 15);
		printf("preorder_traversal:");
		preorder_traversal(tree);
		printf("\npostorder_traversal:");
		postorder_traversal(tree);
		printf("\ninorder_traversal:");
		inorder_traversal(tree);
		printf("\nlevelorder_traversal:");
		levelorder_traversal(tree);
		printf("\n");
		printf("Height: %d, node: %d depth of 15: %d isbinarytree(tree): %d, %d\n", find_height(tree), find_node(tree), find_depth(tree, 15), isbinarytree(tree), isbinarytree2(tree));
		return(0);
}

File: Tree/BST/bst.py
class Tnode(object):
    def __init__(self, left = None, right = None, value = None):
        super().__init__()
        self.left = left
        self.right = right
        self.value = value


class BST(object):
    def __init__(self):
        super().__init__()
        self.root = None

    def insert(self, val):
        new_node = Tnode(value=val)
        if (self.root == None):
            self.root = new_node
        else:
            curr = self.root
            while curr != None:
                if val == curr.value:
                    raise ValueError
                elif val < curr.value:
                    if curr.left == None:
                        curr.left = new_node
                        break
                    curr = curr.left
                else:
                    if curr.right == None:
                        curr.right = new_node
                        break
                    curr = curr.right

    def search(self, val):
        if self.root == None:
            return None
        parent = None
        curr = self.root
        while curr != None:
            if val == curr.value:
                return curr, parent
            elif val < curr.value:
                parent = curr
                curr = curr.left
            else:
                parent = curr
                curr = curr.right
        return None

    def successor(self, node):
        """ node pass to successor must have right child"""
        grandparent = node
        parent = node
        node = node.right
        assert node != None
        while node != None:
            parent = node
            node = node.left
            if node != None:
                grandparent = parent
        return parent, grandparent


    def delete(self, val):
        result = self.search(val)
        if result == None:
            return None
        delnode, parent = result[0], result[1]
        ret_val = delnode.value
        if delnode.left == None and delnode.right == None:
            if delnode == parent.left:
                parent.left = None
            else:
                parent.right = None
        elif not (delnode.left != None and delnode.right != None): # one child
            if delnode.left != None:
                if delnode == parent.left:
                    parent.left = delnode.left
                else:
                    parent.right = delnode.left
            elif delnode == parent.left:  ## delnode.right == None
                parent.left = delnode.left
            else: ## delnode.right == None and delnode == parent.right
                parent.right = delnode.right
        else: ## two child
            succ, par_succ = self.successor(delnode) ## susccessor must not have left child
            delnode.value = succ.value
            if succ.right == None: ## successor have no child
                if succ == par_succ.left:
                    par_succ.left = None
                else:
                    par_succ.right = None
            else: ## successor have one right child
                if succ == par_succ.left:
                    par_succ.left = succ.right
                else:
                    par_succ.right = succ.right
        return ret_val

    def preorder_traversal(self, outnode=None):
        def helper(node):
            if node != None:
                print("%d\t" % (node.value, ), end="")
                helper(node.left)
                helper(node.right)
        print("preorder_traversal:\t", end="")
        if outnode == None:
            helper(self.root)
        else:
            helper(outnode)
        print("")

    def postorder_traversal(self, outnode=None):
        def helper(node):
            if node != None:
                helper(node.left)
                helper(node.right)
                print("%d\t" % (node.value, ), end="")
        print("postorder_traversal:\t", end="")
        if outnode == None:
            helper(self.root)
        else:
            helper(outnode)
        print("")

    def inorder_traversal(self, outnode=None):
        def helper(node):
            if node != None:
                helper(node.left)
                print("%d\t" % (node.value,), end="")
                helper(node.right)
        print("inorder_traversal:\t", end="")
        if outnode == None:
            helper(self.root)
        else:
            helper(outnode)
        print("")

    def levelorder_traversal(self, outnode=None):
        if outnode == None:
            queue = [self.root]
        else:
            queue = [outnode]
        print("levelorder_traversal:\t", end="")
        while (len(queue) != 0):
            node = queue.pop(0)
            print("%d\t" % (node.value,), end="")
            if (node.left != None):
                queue.append(node.left)
            if (node.right != None):
                queue.append(node.right)
        print("")

    def good(self):
        print(self)

    def height(self, outnode=None):
        def helper(node):
            if node == None:
                return -1
            return max(helper(node.left), helper(node.right)) + 1
        if outnode == None:
            outnode = self.root
        return helper(outnode)

    def depth(self, value):
        node = self.root
        dep = 0
        while node != None:
            if value == node.value:
                return dep
            elif value < node.value:
                node = node.left
                dep += 1
            elif value > node.value:
                node = node.right
                dep += 1
        return None

    def node_count(self, outnode=None):
        return self.height(outnode) - 1

    def isbinarytree(self, outnode=None):
        def check_node(node):
            nonlocal pre
            if node == None:
                return True
            elif pre < node.value:
                pre = node.value
                #print(pre)
                return True
            else:
                return False
        def helper(node):
            if node != None:
                if helper(node.left):
                    if check_node(node):
                        if helper(node.right):
                            return True
                return False
            return True

        pre = float("-inf")
        node = self.root if outnode == None else outnode
        return helper(node)




if __name__ == "__main__":
    BSTtree = BST()
    BSTtree.insert(5)
    BSTtree.insert(10)
    BSTtree.insert(15)
    BSTtree.insert(3)
    BSTtree.insert(2)
    BSTtree.insert(1)
    BSTtree.delete(1)
    BSTtree.delete(15)
    BSTtree.insert(1)
    BSTtree.insert(15)
    BSTtree.preorder_traversal()
    BSTtree.postorder_traversal()
    BSTtree.inorder_traversal()
    BSTtree.levelorder_traversal()
    print("BST.height(): %d\t BST.node_count(): %d\t BST.depth(15): %d BST.isbinarytree(): %d" % (BSTtree.height(), BSTtree.node_count(),BSTtree.depth(15), BSTtree.isbinarytree()))
